File: 49. Group Anagrams.py

'''
Given an array of strings strs, group the anagrams together. You can return the answer in any order.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

  I: array of strings
  O: array of arrays
     -each subarray is an array of the same anagram
  C: no duplicates
  E:

  Strategy:
    - create an anagrams object where each key is an alphabetically sorted string, and each value is an array
    - loop through strings and push onto array
    - at the end, map object back into array of arrays



'''
class Solution:
  def groupAnagrams(strs):
    anagrams = {}

    # map strings into anagram
    for string in strs:
      key = [l for l in string]
      key.sort()
      key = ''.join(key)

      try:
        anagrams[key].append(string)
      except:
        anagrams[key] = [string]


    array = []
    # after we've grouped anagrams, map back into an array
    for i, key in enumerate(anagrams):
      array.append(anagrams[key])

    return array



print(Solution.groupAnagrams(["eat","tea","tan","ate","nat","bat"]))

File: 49. Group Anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */

/*  I: array of strings
    O: array of arrays
        -each array is an array of the same anagram
    C: no duplicates
    E:
*/
var groupAnagrams = function(strs) {

  // Tracker object
  let tracker = {}
  // key: alphabetically sorted anagram
  // value: [anagrams]

  // loop through array
  for (var i = 0; i < strs.length; i++) {
      // split each word into array
      let word = strs[i]

      // first sort alphabetically
      let sortedWord = word.split('').sort()

      // paste back together
      sortedWord = sortedWord.join('')

      // if tracker[key] !== undefined -> push onto tracker object key
      if (tracker[sortedWord] !== undefined) {
          tracker[sortedWord].push(word)
      } else {
          // else [value]
          tracker[sortedWord] = [word]
      }

  }


  // Map into outputArray
  let outputArray = []
  // tracker.keys() --> return values into outputArray
  Object.values(tracker).forEach((value) => {
      outputArray.push(value)
  })

  //return outputArray
  return outputArray

};